File: src/romanize.rs
use clap::ValueEnum;
use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Default, PartialEq, Eq, ValueEnum, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RomanizeMode {
    #[default]
    Off,
    Inline,
    Duplicate,
    CurrentOnly,
}

#[derive(Clone, Copy, Default, PartialEq, Eq, ValueEnum, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RomanizeLang {
    /// Chinese pinyin
    Zh,
    /// Japanese romaji (kanji + kana)
    Ja,
    /// Korean revised romanization
    Ko,
    /// Auto-detect per character (Chinese reading for kanji)
    #[default]
    Auto,
}

fn is_cjk(c: char) -> bool {
    matches!(c,
        '\u{4E00}'..='\u{9FFF}'   // CJK Unified Ideographs
        | '\u{3400}'..='\u{4DBF}' // CJK Extension A
        | '\u{F900}'..='\u{FAFF}' // CJK Compatibility
        | '\u{3040}'..='\u{309F}' // Hiragana
        | '\u{30A0}'..='\u{30FF}' // Katakana
        | '\u{AC00}'..='\u{D7AF}' // Hangul Syllables
    )
}

pub fn has_romanizable(text: &str) -> bool {
    text.chars().any(is_cjk)
}

pub fn romanize(text: &str, lang: RomanizeLang) -> String {
    match lang {
        RomanizeLang::Ja => romanize_ja(text),
        RomanizeLang::Zh => romanize_zh(text),
        RomanizeLang::Ko => romanize_generic(text),
        RomanizeLang::Auto => romanize_generic(text),
    }
}

/// Japanese: use kakasi for kanji+kana→romaji
fn romanize_ja(text: &str) -> String {
    kakasi::convert(text).romaji
}

/// Chinese: use pinyin crate for accurate pinyin
fn romanize_zh(text: &str) -> String {
    use pinyin::ToPinyin;

    let mut result = String::new();
    let mut prev_was_pinyin = false;

    for c in text.chars() {
        if let Some(py) = c.to_pinyin() {
            if prev_was_pinyin {
                result.push(' ');
            }
            result.push_str(py.plain());
            prev_was_pinyin = true;
        } else if is_cjk(c) {
            // CJK char not in pinyin dict (kana/hangul) — fallback
            if prev_was_pinyin {
                result.push(' ');
            }
            result.push_str(&any_ascii::any_ascii_char(c));
            prev_was_pinyin = true;
        } else {
            result.push(c);
            prev_was_pinyin = false;
        }
    }
    result
}

/// Generic: any_ascii for Korean hangul and auto mode
fn romanize_generic(text: &str) -> String {
    let mut result = String::new();
    let mut prev_romanized = false;

    for c in text.chars() {
        if is_cjk(c) {
            if prev_romanized {
                result.push(' ');
            }
            result.push_str(&any_ascii::any_ascii_char(c));
            prev_romanized = true;
        } else {
            result.push(c);
            prev_romanized = false;
        }
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    // --- has_romanizable ---

    #[test]
    fn test_has_romanizable_chinese() {
        assert!(has_romanizable("你好世界"));
        assert!(has_romanizable("hello 你好"));
    }

    #[test]
    fn test_has_romanizable_japanese_kana() {
        assert!(has_romanizable("ありがとう"));
        assert!(has_romanizable("カタカナ"));
    }

    #[test]
    fn test_has_romanizable_korean() {
        assert!(has_romanizable("한글"));
    }

    #[test]
    fn test_has_romanizable_latin_only() {
        assert!(!has_romanizable("hello world"));
        assert!(!has_romanizable(""));
        assert!(!has_romanizable("123 !@#"));
    }

    // --- Chinese ---

    #[test]
    fn test_zh_basic() {
        let result = romanize("你好", RomanizeLang::Zh);
        assert!(!result.contains('你'));
        assert!(!result.contains('好'));
        // Should have a space between two pinyin
        assert!(result.contains(' '));
    }

    #[test]
    fn test_zh_mixed() {
        let result = romanize("I love 你", RomanizeLang::Zh);
        assert!(result.starts_with("I love "));
        assert!(!result.contains('你'));
    }

    #[test]
    fn test_zh_preserves_latin() {
        assert_eq!(romanize("hello", RomanizeLang::Zh), "hello");
    }

    // --- Japanese ---

    #[test]
    fn test_ja_hiragana() {
        let result = romanize("ありがとう", RomanizeLang::Ja);
        // Should produce romaji without any kana
        assert!(!result.chars().any(|c| matches!(c, '\u{3040}'..='\u{309F}')));
        assert!(!result.is_empty());
    }

    #[test]
    fn test_ja_katakana() {
        let result = romanize("カタカナ", RomanizeLang::Ja);
        assert!(!result.chars().any(|c| matches!(c, '\u{30A0}'..='\u{30FF}')));
        assert!(!result.is_empty());
    }

    #[test]
    fn test_ja_kanji() {
        let result = romanize("食べる", RomanizeLang::Ja);
        // Should romanize kanji, not just leave it
        assert!(!result.contains('食'));
        assert!(!result.is_empty());
    }

    #[test]
    fn test_ja_preserves_latin() {
        let result = romanize("hello world", RomanizeLang::Ja);
        assert_eq!(result, "hello world");
    }

    // --- Korean ---

    #[test]
    fn test_ko_hangul() {
        let result = romanize("한글", RomanizeLang::Ko);
        assert!(!result.contains('한'));
        assert!(!result.contains('글'));
    }

    // --- Auto ---

    #[test]
    fn test_auto_preserves_latin() {
        assert_eq!(romanize("hello", RomanizeLang::Auto), "hello");
    }

    #[test]
    fn test_auto_romanizes_cjk() {
        let result = romanize("你好", RomanizeLang::Auto);
        assert!(!result.contains('你'));
    }

    // --- No leading/trailing spaces ---

    #[test]
    fn test_no_leading_space() {
        assert!(!romanize("你好", RomanizeLang::Zh).starts_with(' '));
        assert!(!romanize("ありがとう", RomanizeLang::Ja).starts_with(' '));
        assert!(!romanize("한글", RomanizeLang::Ko).starts_with(' '));
    }
}

File: src/romanize.ts
import { pinyin } from "pinyin-pro";
import anyAscii from "any-ascii";

export type RomanizeMode = "off" | "inline" | "duplicate" | "current-only";
export type RomanizeLang = "zh" | "ja" | "ko" | "auto";

/// CJK ranges identical to the Rust `is_cjk`.
function isCjk(c: string): boolean {
  const cp = c.codePointAt(0);
  if (cp === undefined) {
    return false;
  }
  return (
    (cp >= 0x4e00 && cp <= 0x9fff) || // CJK Unified Ideographs
    (cp >= 0x3400 && cp <= 0x4dbf) || // CJK Extension A
    (cp >= 0xf900 && cp <= 0xfaff) || // CJK Compatibility
    (cp >= 0x3040 && cp <= 0x309f) || // Hiragana
    (cp >= 0x30a0 && cp <= 0x30ff) || // Katakana
    (cp >= 0xac00 && cp <= 0xd7af) // Hangul Syllables
  );
}

/// Han ideograph ranges — the subset handled via pinyin in `zh` mode.
function isHan(c: string): boolean {
  const cp = c.codePointAt(0);
  if (cp === undefined) {
    return false;
  }
  return (
    (cp >= 0x4e00 && cp <= 0x9fff) ||
    (cp >= 0x3400 && cp <= 0x4dbf) ||
    (cp >= 0xf900 && cp <= 0xfaff)
  );
}

export function hasRomanizable(text: string): boolean {
  for (const c of text) {
    if (isCjk(c)) {
      return true;
    }
  }
  return false;
}

export function romanize(text: string, lang: RomanizeLang): string {
  switch (lang) {
    case "zh":
      return romanizeZh(text);
    case "ja":
      return romanizeJa(text);
    case "ko":
      return romanizeGeneric(text);
    case "auto":
      return romanizeGeneric(text);
  }
}

/// Chinese: pinyin per Han ideograph, any-ascii fallback for other CJK.
function romanizeZh(text: string): string {
  let result = "";
  let prevWasPinyin = false;

  for (const c of text) {
    if (isHan(c)) {
      if (prevWasPinyin) {
        result += " ";
      }
      result += pinyin(c, { toneType: "none" });
      prevWasPinyin = true;
    } else if (isCjk(c)) {
      // CJK char not handled by pinyin (kana/hangul) — fallback.
      if (prevWasPinyin) {
        result += " ";
      }
      result += anyAscii(c);
      prevWasPinyin = true;
    } else {
      result += c;
      prevWasPinyin = false;
    }
  }
  return result;
}

/// Japanese: approximation of Rust's `romanize_ja` (kakasi). A morphological
/// analyzer is not available, so contiguous CJK runs are transliterated together
/// via any-ascii — kana joins without spaces ("ありがとう" -> "arigatou",
/// matching kakasi for the common case); non-CJK text is preserved verbatim.
/// Kanji readings are context-free (any-ascii), so they still differ from kakasi.
function romanizeJa(text: string): string {
  let result = "";
  let run = "";

  for (const c of text) {
    if (isCjk(c)) {
      run += c;
    } else {
      if (run !== "") {
        result += anyAscii(run);
        run = "";
      }
      result += c;
    }
  }
  if (run !== "") {
    result += anyAscii(run);
  }
  return result;
}

/// Generic: any-ascii per CJK char (Korean hangul, auto).
function romanizeGeneric(text: string): string {
  let result = "";
  let prevRomanized = false;

  for (const c of text) {
    if (isCjk(c)) {
      if (prevRomanized) {
        result += " ";
      }
      result += anyAscii(c);
      prevRomanized = true;
    } else {
      result += c;
      prevRomanized = false;
    }
  }
  return result;
}
